### ISSUE TITLE
local *_; goto &sub; segfaults on 5.18

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13302]
Migrated from [rt.perl.org#119949](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=119949) (status was 'resolved')

Searchable as RT119949$

**Comment by p5pRT** [comment]
### From @Hugmeir

#### Created by fraserb@gmail.com



sub foo \{ say "in foo&#8203;: \<@&#8203;\_>" \}
sub bar \{ local \*\_ = "hello foo"; goto &foo \}
bar\("hello bar"\)

This incorrectly prints "in foo&#8203;: \<hello bar>" in 5\.16\, and segfaults in
5\.18 and 5\.19\.4\. It should probably just give an uninit warning and print
"in foo&#8203;: \<>"\.
This is likely a leftover from making 'local @&#8203;\_ = \.\.\.; goto &sub;' work\.
Something in pp\_leavesub is likely making unwarranted assumptions about
GvAV\(PL\_defav\) or cx\->blk\_sub\.argarray\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.18.1:

Configured by hugmeir at Thu Sep 12 14:58:14 ART 2013.

Summary of my perl5 (revision 5 version 18 subversion 1) configuration:

  Platform:
    osname=darwin, osvers=12.4.0, archname=darwin-thread-multi-2level
    uname='darwin airtiger 12.4.0 darwin kernel version 12.4.0: wed may 1
17:57:12 pdt 2013; root:xnu-2050.24.15~1release_x86_64 x86_64 '
    config_args='-de
-Dprefix=/Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1 -Dusethreads
-DDEBUGGING
-Aeval:scriptdir=/Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector',
    optimize='-O3 -g',
    cppflags='-fno-common -DPERL_DARWIN -DDEBUGGING -fno-strict-aliasing
-pipe -fstack-protector'
    ccversion='', gccversion='4.2.1 Compatible Apple LLVM 4.2
(clang-425.0.28)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector'
    libpth=/usr/lib
    libs=-ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup
-fstack-protector'

Locally applied patches:



@INC for perl 5.18.1:
    /Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
/lib/perl5/darwin-thread-multi-2level
    /Users/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5

/Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/lib/site_perl/5.18.1/darwin-thread-multi-2level
    /Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/lib/site_perl/5.18.1

/Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/lib/5.18.1/darwin-thread-multi-2level
    /Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/lib/5.18.1
    .


Environment for perl 5.18.1:
    DYLD_LIBRARY_PATH=/usr/local/mysql/lib:
    HOME=/Users/hugmeir
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=/Users/hugmeir/perl5/perlbrew/bin:/Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/local/git/bin:/usr/local/mysql/bin
    PERL5LIB=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5
    PERLBREW_BASHRC_VERSION=0.58
    PERLBREW_HOME=/Users/hugmeir/.perlbrew
    PERLBREW_LIB=all
    PERLBREW_MANPATH=/Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/man

PERLBREW_PATH=/Users/hugmeir/perl5/perlbrew/bin:/Users/hugmeir/perl5/perlbrew/perls/perl-5.18.1/bin
    PERLBREW_PERL=perl-5.18.1
    PERLBREW_ROOT=/Users/hugmeir/perl5/perlbrew
    PERLBREW_VERSION=0.58
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MB_OPT=--install_base /Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MM_OPT=INSTALL_BASE=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Sep 22\, 2013 at 08&#8203;:50&#8203;:10PM \-0700\, Brian Fraser wrote&#8203;:

> sub foo \{ say "in foo&#8203;: \<@&#8203;\_>" \}
> sub bar \{ local \*\_ = "hello foo"; goto &foo \}
> bar\("hello bar"\)
> 
> This incorrectly prints "in foo&#8203;: \<hello bar>" in 5\.16\, and segfaults in
> 5\.18 and 5\.19\.4\. It should probably just give an uninit warning and print
> "in foo&#8203;: \<>"\.
> This is likely a leftover from making 'local @&#8203;\_ = \.\.\.; goto &sub;' work\.
> Something in pp\_leavesub is likely making unwarranted assumptions about
> GvAV\(PL\_defav\) or cx\->blk\_sub\.argarray\.

Indeed\,

$ bisect\.pl \-\-target miniperl \-\-start v5\.16\.0 \-le 'sub foo \{ print "in foo&#8203;: \<@&#8203;\_>" \} sub bar \{ local \*\_ = "hello foo"; goto &foo \} bar\("hello bar"\)'


says

049bd5ffd62b73325d4b2e75e59ba04b3569137d is the first bad commit
commit 049bd5ffd62b73325d4b2e75e59ba04b3569137d
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Sun Nov 11 22&#8203;:16&#8203;:35 2012 \-0800

&nbsp;    \[perl \#43077\] Make goto &sub leave @&#8203;\_ alone

&nbsp;    It is a little tricky\, as we have to hang on to @&#8203;\_ while unwinding the
&nbsp;    effects of local @&#8203;\_\.


and at that revision behaviour changes from "in foo&#8203;: \<hello bar>\\n" to SEGV\.


Program received signal SIGSEGV\, Segmentation fault\.
0x00000000004ce904 in Perl\_pp\_leavesub \(\) at pp\_hot\.c&#8203;:2604
2604        POPSUB\(cx\,sv\);      /\* Stack values are safe&#8203;: release CV and @&#8203;\_ \.\.\. \*/

valgrind said

==23520==  Address 0xc is not stack'd\, malloc'd or \(recently\) free'd

So it's a structure offset of 12 from a NULL pointer\.

\(gdb\) p cx\->cx\_u\.cx\_blk\.blk\_u\.blku\_sub
$5 = \{retop = 0x7e5c50\, cv = 0x7ddd00\, savearray = 0x7caf38\, argarray = 0x0\,
&nbsp;  olddepth = 0\, oldcomppad = 0x7ddd18\}

argarray is NULL\.


I don't understand enough of this to fix it without working it out from
"first principals"\.

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Sep 23 01&#8203;:40&#8203;:58 2013\, nicholas wrote&#8203;:
> On Sun\, Sep 22\, 2013 at 08&#8203;:50&#8203;:10PM \-0700\, Brian Fraser wrote&#8203;:
> 
> > sub foo \{ say "in foo&#8203;: \<@&#8203;\_>" \}
> > sub bar \{ local \*\_ = "hello foo"; goto &foo \}
> > bar\("hello bar"\)
> >
> > This incorrectly prints "in foo&#8203;: \<hello bar>" in 5\.16\, and segfaults
> in
> > 5\.18 and 5\.19\.4\. It should probably just give an uninit warning and
> print
> > "in foo&#8203;: \<>"\.
> > This is likely a leftover from making 'local @&#8203;\_ = \.\.\.; goto &sub;'
> work\.
> > Something in pp\_leavesub is likely making unwarranted assumptions
> about
> > GvAV\(PL\_defav\) or cx\->blk\_sub\.argarray\.
> 
> Indeed\,
> 
> $ bisect\.pl \-\-target miniperl \-\-start v5\.16\.0 \-le 'sub foo \{ print "in
> foo&#8203;: \<@&#8203;\_>" \} sub bar \{ local \*\_ = "hello foo"; goto &foo \} bar\("hello
> bar"\)'
> 
> 
> says
> 
> 049bd5ffd62b73325d4b2e75e59ba04b3569137d is the first bad commit
> commit 049bd5ffd62b73325d4b2e75e59ba04b3569137d
> Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> Date&#8203;:   Sun Nov 11 22&#8203;:16&#8203;:35 2012 \-0800
> 
>     \[perl \#43077\] Make goto &sub leave @&#8203;\_ alone
> 
>     It is a little tricky\, as we have to hang on to @&#8203;\_ while unwinding
> the
>     effects of local @&#8203;\_\.
> 
> 
> and at that revision behaviour changes from "in foo&#8203;: \<hello bar>\\n" to
> SEGV\.
> 
> 
> Program received signal SIGSEGV\, Segmentation fault\.
> 0x00000000004ce904 in Perl\_pp\_leavesub \(\) at pp\_hot\.c&#8203;:2604
> 2604        POPSUB\(cx\,sv\);      /\* Stack values are safe&#8203;: release CV
> and @&#8203;\_ \.\.\. \*/
> 
> valgrind said
> 
> ==23520==  Address 0xc is not stack'd\, malloc'd or \(recently\) free'd
> 
> So it's a structure offset of 12 from a NULL pointer\.
> 
> \(gdb\) p cx\->cx\_u\.cx\_blk\.blk\_u\.blku\_sub
> $5 = \{retop = 0x7e5c50\, cv = 0x7ddd00\, savearray = 0x7caf38\, argarray
> = 0x0\,
>   olddepth = 0\, oldcomppad = 0x7ddd18\}
> 
> argarray is NULL\.

It looks as though we need to changes things here\, in pp\_goto&#8203;:

&nbsp;        if \(CxHASARGS\(cx\)\)
&nbsp;        \{
&nbsp;            CX\_CURPAD\_SAVE\(cx\->blk\_sub\);

&nbsp;            /\* cx\->blk\_sub\.argarray has no reference count\, so we
&nbsp;               need something to hang on to our argument array so
&nbsp;               that cx\->blk\_sub\.argarray does not end up pointing
&nbsp;               to freed memory as the result of undef \*\_\.  So put
&nbsp;               it in the callee’s pad\, donating our refer\-
&nbsp;               ence count\. \*/
&nbsp;            SvREFCNT\_dec\(PAD\_SVl\(0\)\);
&nbsp;            PAD\_SVl\(0\) = \(SV \*\)\(cx\->blk\_sub\.argarray = arg\);

&nbsp;            /\* GvAV\(PL\_defgv\) might have been modified on scope
&nbsp;               exit\, so restore it\. \*/
&nbsp;            if \(arg \!= GvAV\(PL\_defgv\)\) \{
&nbsp;            AV \* const av = GvAV\(PL\_defgv\);
&nbsp;            GvAV\(PL\_defgv\) = \(AV \*\)SvREFCNT\_inc\_simple\(arg\);
&nbsp;            SvREFCNT\_dec\(av\);
&nbsp;            \}
&nbsp;        \}
&nbsp;        else SvREFCNT\_dec\(arg\);

If ‘arg’ is null here\, we probably ought to turn off CxHASARGS and
localise GvAV\(PL\_defgv\) using the savestack\.

Sorry\, I can’t do this right now\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

This is now fixed in bfa371b\.

Can we backport this to 5\.18?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2014\-01\-19T00&#8203;:33&#8203;:24\]
> This is now fixed in bfa371b\.
> 
> Can we backport this to 5\.18?

\+1\, and I see it's already on the 5\.18\.3 ticket\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Jan 20 05&#8203;:02&#8203;:17 2014\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2014\-01\-
> 19T00&#8203;:33&#8203;:24\]
> > This is now fixed in bfa371b\.
> >
> > Can we backport this to 5\.18?
> 
> \+1\, and I see it's already on the 5\.18\.3 ticket\.


\+1 from me\, which is a third vote\, so cherry\-picked with some noise as 2c60386e935e57af3d9202ad4b798662d2e997a1\.

perldelta update in b380e583f8ebe423cffa9f4986cf213064b3eb59\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
